jib-core/src/test/java/com/google/cloud/tools/jib/http/TestWebServer.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2018 Google LLC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.http;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+
+/** Simple local web server for testing. */
+class TestWebServer implements Closeable {
+
+  private final ServerSocket serverSocket;
+  private final List<Socket> socketsToClose = new ArrayList<>();
+  private final ExecutorService executorService = Executors.newSingleThreadExecutor();
+  private final Semaphore threadStarted = new Semaphore(0);
+
+  TestWebServer() throws IOException, InterruptedException {
+    serverSocket = new ServerSocket(0);
+    executorService.submit(this::serve200);
+    threadStarted.acquire();
+  }
+
+  String getEndpoint() {
+    String host = serverSocket.getInetAddress().getHostAddress();
+    return "http://" + host + ":" + serverSocket.getLocalPort();
+  }
+
+  @Override
+  public void close() throws IOException {
+    serverSocket.close();
+    for (Socket socket : socketsToClose) {
+      try (Socket toClose = socket) {
+      } catch (IOException ex) {
+      }
+    }
+    executorService.shutdown();
+  }
+
+  private Void serve200() throws IOException {
+    threadStarted.release();
+    while (true) {
+      Socket socket = serverSocket.accept();
+      socketsToClose.add(socket);
```

Comment:
It's safe to call `close()` multiple times on a socket or call it on a closed socket. Probably it is safe on almost all `Closeable` classes.

(And I don't think I get the out-of-bound exception. The enhanced `for` loop never throws it.)

Reply:
Oh sorry I was referring to `TestWebServer#close`, where it loops over the `socketsToClose` non-concurrent list, but this list is being updated (via `add`) in another thread. Perhaps the `executorService.shutdown()` should be called at the start of `TestWebServer#close`.

Reply:
Ah, haven't thought about that. Using `ConcurrentLinkedQueue` would make it safe. If `close()` is called concurrently, it may still miss a socket from a running thread, but that's OK.

BTW, I'd like to give a few more final tests on Windows.

Reply:
That's a very nice idea.

Reply:
Another idea that can close sockets early would be to wait until the client has closed the channel (i.e., checking `InputStream` is still open). However, the current implementation seems to work and enough for testing.

Reply:
> wait until the client has closed the channel

Yup, it works. This is definitely the right way. No more weird lazy close, and no need to flush either.